### PR TITLE
Cache

### PR DESCRIPTION
**Pull request by chrystalquek**
- [x] Necessity of UserInfo [follow recipe more standard way]
- [x] Sprinkle completion function in more places and check accuracy
- [ ] Large volume boundary testing (done for users)
- [x] Batched writes across functions called in StorageManager [not much added benefit]
- [ ] Remove CombineFirebase library
- [x] Defining strings e.g. userId maps to "creator" in a meaningful place [ignore]
- [x] Updating tests
- [x] Completion function signature - should be (..?, ..?) cos may want to pass the Error up for display
- [x] Ignoring alot of errors now (eg in batch commit etc)
- [x] Rename functions (remove 'byId' etc)
- [x] Refresh button (ignore: just switch tabs)
- [x] imageUpdatedAt field in OnlineRecipeInfoRecord - https://developer.apple.com/documentation/foundation/nsdata/1409330-isequal
- [ ] Pagination to deal with 10 arr limit [do l8ter]
- [x] Deal with OnlineRecipe deletion: Use OnlineRecipeCollectionEditor
- [ ] Many imports of UIImage into StorageManager/Image related stuff
- [x] Change some stuffs to structs
- [x] Things are not atomic (ignore)

memory leak??

**Problems that I'm not sure whether I should solve**
1. Current: Invoke entire list loading on add or delete, invoke single entity loading on edit. 
Why this is a problem: May be unnecessarily invoking entire list loading on add/delete. Also need to pass some functions from parent view model to child view models (take a look at `reload`)
Things that could be done
- ChildViewModels maintain a reference to ParentViewModels OR Use some event thingy + instantiate child view models in parent view models [both r very painful solutions]
- Get back entity from firebase after add (can't directly add from entity that i have because lack information such as timestamps)

[instead: create common object that both parent and child view models observe - check editorGraph]

- Common Object

2. Singleton cache
Why this is a problem: Singletons not encouraged
Why we may not want to solve it: AppDatabase is also a singleton, more consistency. Hard to make it non singleton.

3. Many imports of UIImage into StorageManager/Image related stuff

**Problems that I cannot solve**
1. FirebaseDatabase has knowledge of RunTime models, instantiates them from records and stores them in the cache. FirebaseDatabase needs to have knowledge of the cache to know which stuff to fetch
2. Tight knitting of cache (alot of coupling now, v hard to remove if say don't want to use cache in the future)

### CODE REVIEW

File: ChopChop/Models/User/User.swift [hunk 1]
```diff
@@ -1,28 +1,42 @@
...
+        self.updatedAt = updatedAt
     }
+
 }
 
-extension User: Codable {
+extension User {
+    init?(from record: UserRecord, infoRecord: UserInfoRecord) {
+        guard let id = record.id, let createdAt = infoRecord.createdAt, let updatedAt = infoRecord.updatedAt else {
+            return nil // or throw errors?
```
Comment (sc-arecrow) [inline ChopChop/Models/User/User.swift:30]: I think it depends on when the optional fields are `nil`

File: ChopChop/Storage/Cache.swift [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            // then we remove any value for that key:
+                removeValue(forKey: key)
+                return
+            }
+
+            insert(value, forKey: key)
+        }
+    }
+
+    // NOTE: No type safety here because can compare any idss
```
Comment (sc-arecrow) [inline ChopChop/Storage/Cache.swift]: Since the type of `Value` is generic and restricted to the `CachableEntity` protocol, would this still be an issue?

File: ChopChop/Storage/ChopChopCache.swift [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import Foundation
+
+class ChopChopCache {
+
+    let onlineRecipeCache = Cache<String, OnlineRecipe>()
+    let userCache = Cache<String, User>()
+    var onlineRecipeImageCache = [String: Data]() // not using cache because Data cannot conform to CachableEntity
```
Comment (sc-arecrow) [inline ChopChop/Storage/ChopChopCache.swift]: Is it possible to create a wrapper around `Data` to make it conform?

File: ChopChop/Storage/ChopChopCache.swift [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import Foundation
+
+class ChopChopCache {
+
+    let onlineRecipeCache = Cache<String, OnlineRecipe>()
+    let userCache = Cache<String, User>()
```
Comment (sc-arecrow) [inline ChopChop/Storage/ChopChopCache.swift]: How are the keys generated?

File: ChopChop/Storage/Models/OnlineRecipes/OnlineRecipeInfoRecord.swift [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import Foundation
+import FirebaseFirestoreSwift
+
+struct OnlineRecipeInfoRecord: InfoRecord {
+    @DocumentID var id: String? // same as OnlineRecipe id
+    private(set) var creator: String // creatorId
```
Comment (sc-arecrow) [inline ChopChop/Storage/Models/OnlineRecipes/OnlineRecipeInfoRecord.swift]: Perhaps just name this property `creatorId` instead

File: ChopChop/Storage/StorageManager.swift [hunk 6]
```diff
@@ -216,16 +219,29 @@ extension StorageManager {
...
     func fetchRecipeImage(name: String) -> UIImage? {
         ImageStore.fetch(imageNamed: name, inFolderNamed: StorageManager.recipeFolderName)
     }
 
-    func saveRecipeImage(_ image: UIImage, name: String) throws {
+    func saveRecipeImage(_ image: UIImage, id: Int64, name: String) throws {
         do {
+            let originalImage = fetchRecipeImage(name: name)
+            let isSameImage = originalImage?.pngData() == image.pngData()
+            guard var recipe = try fetchRecipe(id: id), !isSameImage else {
```
Comment (sc-arecrow) [inline ChopChop/Storage/StorageManager.swift]: Checking for same image before fetching recipe might improve performance

File: ChopChop/ViewModels/CreateUserProfileViewModel.swift [hunk 7]
```diff
@@ -19,17 +19,19 @@ final class CreateUserProfileViewModel: ObservableObject {
...
-            assertionFailure()
-            return
+        let userId = try? storageManager.addUser(name: name) { err in
+            if err != nil {
+                assertionFailure()
+                return
+            }
         }
 
-        UserDefaults.standard.set(userId, forKey: "userId")
```
Comment (sc-arecrow) [inline ChopChop/ViewModels/CreateUserProfileViewModel.swift:32]: Why was this removed?

File: ChopChop/ViewModels/Users/ProfileViewModel.swift [hunk 8]
```diff
@@ -7,46 +7,28 @@ import Combine
...
-    /// The ids of the followees of the user.
-    private var followeeIds: [String] = []
+
+    private let storageManager = StorageManager()
+    private let settings: UserSettings
+    @ObservedObject private(set) var recipesViewModel: OnlineRecipeCollectionViewModel
 
     /// User profile details
-    @Published private(set) var userName = ""
+    @Published private(set) var userName = "No name"
```
Comment (sc-arecrow) [inline ChopChop/ViewModels/Users/ProfileViewModel.swift]: Should we change all the "No name"s to empty strings instead? Because users can't have empty strings as names, but they could name themselves "no name"

File: ChopChop/Views/OnlineRecipes/OnlineRecipeBySelfView.swift [hunk 9]
```diff
@@ -1,7 +1,7 @@
 import SwiftUI
 
 struct OnlineRecipeBySelfView: View {
-    @ObservedObject var viewModel: OnlineRecipeBySelfViewModel
+    let viewModel: OnlineRecipeBySelfViewModel
```
Comment (sc-arecrow) [inline ChopChop/Views/OnlineRecipes/OnlineRecipeBySelfView.swift:4]: If the view model is no longer observed, does `OnlineRecipeViewModel` still need to be an observable object?

File: ChopChop/Storage/Cache.swift (hunk 2 above)
Comment (chrystalquek) [reply ChopChop/Storage/Cache.swift]: I was thinking along the lines of the fact that i can get a userInfoRecord, use its `id` and `updatedAt` to call the `getEntityIfCachedAndValid` method of the onlineRecipeCache, and get an OnlineRecipe. Which is abit strange. but not entirely sure whether this is considered an issue

But if the ids had the type safety stuff, this would not have been an issue

File: ChopChop/Storage/ChopChopCache.swift (hunk 3 above)
Comment (chrystalquek) [reply ChopChop/Storage/ChopChopCache.swift]: will change

File: ChopChop/Storage/ChopChopCache.swift (hunk 4 above)
Comment (chrystalquek) [reply ChopChop/Storage/ChopChopCache.swift]: keys are the ids in firebase

File: ChopChop/Storage/Models/OnlineRecipes/OnlineRecipeInfoRecord.swift (hunk 5 above)
Comment (chrystalquek) [reply ChopChop/Storage/Models/OnlineRecipes/OnlineRecipeInfoRecord.swift]: will change

File: ChopChop/ViewModels/CreateUserProfileViewModel.swift (hunk 7 above)
Comment (chrystalquek) [reply ChopChop/ViewModels/CreateUserProfileViewModel.swift:32]: Realised that its already set when i update userId of UserSettings so no longer necessary

File: ChopChop/ViewModels/Users/ProfileViewModel.swift (hunk 8 above)
Comment (chrystalquek) [reply ChopChop/ViewModels/Users/ProfileViewModel.swift]: Will change

File: ChopChop/Views/OnlineRecipes/OnlineRecipeBySelfView.swift (hunk 9 above)
Comment (chrystalquek) [reply ChopChop/Views/OnlineRecipes/OnlineRecipeBySelfView.swift:4]: In this particular view, OnlineRecipeBySelfViewModel does not need to be observed because it does not have any @Published properties.

OnlineRecipeViewModel still needs to be an observable object because another view, ie OnlineRecipeView, needs to observe OnlineRecipeViewModel

File: ChopChop/Models/Recipes/Recipe.swift [hunk 10]
```diff
@@ -4,6 +4,7 @@ import GRDB
 struct Recipe: Equatable {
     var id: Int64?
     var onlineId: String?
+    var isImageUploaded: Bool?
```
Comment (seowalex) [inline ChopChop/Models/Recipes/Recipe.swift]: Why is this a optional field? Shouldn't there be only two states for whether an image is uploaded?

File: ChopChop/Storage/StorageManager.swift [hunk 11]
```diff
@@ -216,16 +219,29 @@ extension StorageManager {
     static let ingredientFolderName = "Ingredient"
     static let recipeFolderName = "Recipe"
 
-    func deleteRecipeImage(name: String) {
+    func deleteRecipeImage(name: String, id: Int64) throws {
+        guard var recipe = try fetchRecipe(id: id) else {
+            return
+        }
+        recipe.isImageUploaded = nil
+        try saveRecipe(&recipe)
```
Comment (seowalex) [inline ChopChop/Storage/StorageManager.swift]: This should probably be done in `RecipeFormViewModel` instead, or the recipe will be saved twice every time, which might cause lags if the step graph is large.

File: ChopChop/Storage/StorageManager.swift [hunk 12]
```diff
@@ -216,16 +219,29 @@ extension StorageManager {
...
-    func saveRecipeImage(_ image: UIImage, name: String) throws {
+    func saveRecipeImage(_ image: UIImage, id: Int64, name: String) throws {
         do {
+            let originalImage = fetchRecipeImage(name: name)
+            let isSameImage = originalImage?.pngData() == image.pngData()
+            guard var recipe = try fetchRecipe(id: id), !isSameImage else {
+                return
+            }
+            recipe.isImageUploaded = false
+            try saveRecipe(&recipe)
```
Comment (seowalex) [inline ChopChop/Storage/StorageManager.swift]: Again, this might be better done in `RecipeFormViewModel`, especially since it already loads the original image (if it exists).

File: ChopChop/Storage/StorageManager.swift [hunk 13]
```diff
@@ -278,28 +296,31 @@ extension StorageManager {
             steps: nodes,
             stepEdges: edgeRecords
         )
-        let onlineId = try firebase.addRecipe(recipe: recipeRecord)
-        recipe.onlineId = onlineId
-        try self.saveRecipe(&recipe)
 
         guard let id = recipe.id else {
+            assertionFailure("Should have id")
```
Comment (seowalex) [inline ChopChop/Storage/StorageManager.swift]: A side note, but I think we may want to remove any assertion failures unless a failure means that the application itself cannot work (like being unable to access the local database).

File: ChopChop/ViewModels/CreateUserProfileViewModel.swift [hunk 14]
```diff
@@ -19,17 +19,19 @@ final class CreateUserProfileViewModel: ObservableObject {
             return
         }
 
-        guard let newUser = try? User(name: name) else {
+        // Its not possible to create a user with just aname
+        guard !name.isEmpty else {
```
Comment (seowalex) [inline ChopChop/ViewModels/CreateUserProfileViewModel.swift]: Do you need to trim this?

File: ChopChop/ViewModels/Recipes/RecipeViewModel.swift [hunk 15]
```diff
@@ -41,32 +41,30 @@ final class RecipeViewModel: ObservableObject {
...
-                    }
-                }, receiveValue: { value in
-                    self.parentRecipe = value
-                })
+        guard let parentId = recipe?.parentOnlineRecipeId else {
+            return
+        }
+        storageManager.fetchOnlineRecipe(id: parentId) { onlineRecipe, _ in
+            self.parentRecipe = onlineRecipe
         }
```
Comment (seowalex) [inline ChopChop/ViewModels/Recipes/RecipeViewModel.swift:49]: Does this mean that if the network is slow, the recipe view will take a long time to load? It seems like this method isn't asynchronous, unlike the previous publisher.

File: ChopChop/Models/OnlineRecipes/OnlineRecipe.swift [hunk 16]
```diff
@@ -1,24 +1,25 @@
...
-
-    private(set) var name: String
-    private(set) var servings: Double
-    private(set) var cuisine: String?
-    private(set) var difficulty: Difficulty?
-    private(set) var stepGraph: RecipeStepGraph
-    private(set) var ingredients: [RecipeIngredient]
-    private(set) var ratings: [RecipeRating]
-    private(set) var created: Date
+struct OnlineRecipe: Identifiable, CachableEntity {
```
Comment (shadowezz) [inline ChopChop/Models/OnlineRecipes/OnlineRecipe.swift:3]: Hmm why is it a struct now? Is it so that it can conform to cachableEntity? And i checked our report say OnlineRecipe is a class xD

File: ChopChop/Storage/StorageManager.swift (hunk 12 above)
Comment (chrystalquek) [reply ChopChop/Storage/StorageManager.swift]: I can move it to RecipeFormViewModel, but i will still need to reload the original image because it seems that the new image overwrites the image in RecipeFormViewModel (correct me if im wrong)

File: ChopChop/ViewModels/Recipes/RecipeViewModel.swift (hunk 15 above)
Comment (chrystalquek) [reply ChopChop/ViewModels/Recipes/RecipeViewModel.swift:49]: the firebase operations are all asynchronous, so that closure part is like a function that is passed as a param, and only called when firebase finishes the api call

File: ChopChop/Models/OnlineRecipes/OnlineRecipe.swift (hunk 16 above)
Comment (chrystalquek) [reply ChopChop/Models/OnlineRecipes/OnlineRecipe.swift:3]: lol uh just change ah, no point making into a class

File: ChopChop/ViewModels/CreateUserProfileViewModel.swift [hunk 17]
```diff
@@ -19,17 +19,19 @@ final class CreateUserProfileViewModel: ObservableObject {
...
-            return
+        let userId = try? storageManager.addUser(name: name) { err in
+            if err != nil {
+                assertionFailure()
+                return
+            }
         }
 
-        UserDefaults.standard.set(userId, forKey: "userId")
         settings.userId = userId
```
Comment (shadowezz) [inline ChopChop/ViewModels/CreateUserProfileViewModel.swift]: If the batched write fails will this still run?

File: ChopChop/Storage/Cache.swift [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+    func getEntityIfCachedAndValid(id: Key, updatedDate: Date) -> Value? {
+        guard let entity = value(forKey: id), entity.updatedAt >= updatedDate else {
+            return nil
+        }
+        return entity
+    }
+
+}
+
+extension Cache {
```
Comment (shadowezz) [inline ChopChop/Storage/Cache.swift]: Just curious why is there a need for a wrapper around key and value

**Review by shadowezz — COMMENTED** [review]
Some comments

File: ChopChop/Storage/StorageManager.swift (hunk 12 above)
Comment (seowalex) [reply ChopChop/Storage/StorageManager.swift]: But if there is a new image, doesn't that just mean you should change the `isImageUploaded` variable to true in the new recipe that is saved?

File: ChopChop/Storage/Cache.swift (hunk 18 above)
Comment (chrystalquek) [reply ChopChop/Storage/Cache.swift]: NSCache is very hard to use without the wrapper
